src/pudl/analysis/record_linkage/name_cleaner.py

```diff
@@ -133,61 +192,69 @@ def _apply_regex_rules(
 
             # Make sure to use raw string
             regex_rule = rf"{regex_rule}"
-
-            # Treat the special case of the word THE at the end of a text's name
-            found_the_word_the = None
-            if name_rule == "place_word_the_at_the_beginning":
-                found_the_word_the = re.search(regex_rule, clean_value)
-
             # Apply the regex rule
-            clean_value = re.sub(regex_rule, replacement, clean_value)
+            clean_col = clean_col.str.replace(regex_rule, replacement, regex=True)
 
-            # Adjust the name for the case of rule <place_word_the_at_the_beginning>
-            if found_the_word_the is not None:
-                clean_value = "the " + clean_value
+        return clean_col
 
-        return clean_value
-
-    def _remove_unicode_chars(self, value: str) -> str:
-        """Removes unicode character that is unreadable when converted to ASCII format.
+    def _remove_unicode_chars(self, col: pd.Series) -> pd.Series:
+        """Removes unicode characters that are unreadable in ASCII format.
 
         Arguments:
-            value (str): any string containing unicode characters.
+            col (pd.Series): series containing unicode characters.
 
         Returns:
-            (str): the corresponding input string without unicode characters.
+            (pd.Series): the corresponding input series without unicode characters.
         """
-        # Remove all unicode characters if any
-        clean_value = value.encode("ascii", "ignore").decode()
-        return clean_value
+        return col.str.encode("ascii", "ignore").str.decode("ascii")
+
+    def _move_the_to_beginning(self, col: pd.Series) -> pd.Series:
+        remove_the_from_end_regex_rule = CLEANING_RULES_DICT[
+            "remove_word_the_from_the_end"
+        ][1]
+        remove_the_from_end_replacement = CLEANING_RULES_DICT[
+            "remove_word_the_from_the_end"
+        ][0]
+        # find matches with the at end
+        the_at_end_matches = col.str.contains(
+            remove_the_from_end_regex_rule, regex=True
+        )
+
+        # remove the from end of strings
+        clean_col = col.str.replace(
+            remove_the_from_end_regex_rule,
+            remove_the_from_end_replacement,
+            regex=True,
+        )
+        clean_col = clean_col.where(~the_at_end_matches, "the " + clean_col)
+        return clean_col
 
-    def _apply_cleaning_rules(self, company_name: str) -> str:
+    def _apply_cleaning_rules(self, col: pd.Series) -> pd.Series:
         """Apply the cleaning rules from the dictionary of regex rules."""
+        if self.place_word_the_at_beginning:
```

Comment:
Why is this rule handled differently? Is it a combination of two rules?

Reply:
This has to do with the relationship between the `"remove_word_the_from_the_end"`, `"remove_word_the_from_the_beginning"`, and `place_word_the_at_beginning` rules. I decided that if you want to place the word the at the beginning, then you should do this first in case you want to then remove "the" from the end or "the" from the beginning. These rules are all kind of in conflict and feed into the idea that there are "states" that we go through - an `op` to handle "the" would make sense in a later refactor.